csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +398,479 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
+                            {
+                                waitTask = stopTask;
+                            }
+                            else if (stopTask.Value.IsFaulted || stopTask.Value.IsCanceled)
+                            {
+                                // Already started, no need to do anything.
+                                _stopTask = null;
+                                _startTask = new(Task.CompletedTask);
+                                performTask = _startTask;
+                            }
+                            else if (!stopTask.Value.IsCompletedSuccessfully)
+                            {
+                                waitTask = stopTask; // we need to wait until this task is complete
+                            }
+                            // else stopTask completed successfully and we replace it with a new startTask below.
+                        }
+                        else if (_startTask is Lazy<Task> startTask)
+                        {
+                            if (!startTask.IsValueCreated ||
+                                startTask.Value.IsCompletedSuccessfully ||
+                                !startTask.Value.IsCompleted)
+                            {
+                                performTask = startTask;
+                            }
+                            // else, startTask is canceled or faulted and we create a new startTask below.
+                        }
+                        else
+                        {
+                            Debug.Assert(false);
+                            performTask = new(Task.CompletedTask); // never reached
+                        }
+
+                        if (waitTask == null && performTask == null)
+                        {
+                            _stopTask = null;
+
+                            // The async/await ensures we get a Task (i.e. successful lazy initialization) even if
+                            // StartAsync throws an exception synchronously.
+                            // Note that StartAsync is not executed immediately with _mutex locked.
+                            _startTask = new(
+                                    async () => await _service.StartAsync(
+                                                    Name,
+                                                    Communicator ?? sharedCommunicator!,
+                                                    _args,
+                                                    cancel).ConfigureAwait(false));
+
+                            performTask = _startTask;
+                        }
                     }
-                    catch (Exception ex)
+
+                    if (waitTask != null)
                     {
-                        _logger.Warning(
-                            $"IceBox.ServiceManager: exception while destroying shared communicator:\n{ex}");
+                        await waitTask.Value.WaitAsync(cancel);
```

Comment:
Shouldn't we ignore exceptions from this waitTask, and return a completed task, seems more consistent with what we do above for `stopTask.Value.IsFaulted || stopTask.Value.IsCanceled`

Reply:
Yes, you're right, fixing.

Reply:
Please re-review.